### PR TITLE
Support fetchOrderBooks for multiple symbols at once #1033

### PR DESCRIPTION
**Pull request by pmizel**
hope this can help. pr is done.

надеюсь все окей.

### CODE REVIEW

**Comment by kroitor** [comment]
Oh-oh... But... why did you delete the spacing? %)

**Comment by pmizel** [comment]
my auto prettifier did this job...sorry.

**Comment by pmizel** [comment]
but build was succesful in my case.

**Comment by kroitor** [comment]
@pmizel can you switch it off and resave them? It will break the build and transpilation otherwise, unfortunately:

https://travis-ci.org/ccxt/ccxt/builds/324642643?utm_source=github_status&utm_medium=notification

screen shot 2018-01-03 at 19 16 39

File: js/ccex.js [hunk 1]
```diff
@@ -131,6 +133,48 @@ module.exports = class ccex extends Exchange {
         return this.parseOrderBook (orderbook, undefined, 'buy', 'sell', 'Rate', 'Quantity');
     }
 
+    async fetchOrderBooks(symbols = undefined, params = {}){
+        await this.loadMarkets ();
+        let orderBooks = [];
+        let orderBooksResult = [];
+
```
Comment (kroitor) [inline js/ccex.js]: Empty lines are not allowed inside methods by the following rules: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes (we really have to follow them all literally).

File: js/ccex.js [hunk 2]
```diff
@@ -131,6 +133,48 @@ module.exports = class ccex extends Exchange {
...
+        try {
+            let response = await this.publicGetFullorderbook();
+            if(response && response.success && response.result) {
+                for (let type of Object.keys(response.result)) {
+                    for (let order of response.result[type]) {
+                        let index = orderBooks.findIndex(function (f){
+                            return f.symbol === order.Market.replace("-", "/").toUpperCase()
```
Comment (kroitor) [inline js/ccex.js]: Assoc keys should be single quoted `f['symbol']` = good, `f.symbol` = bad (won't transpile).

File: js/ccex.js [hunk 3]
```diff
@@ -131,6 +133,48 @@ module.exports = class ccex extends Exchange {
...
+                        }
+
+                        if (type === "buy") {
+                            orderBooks[index].bids.push([order.Rate, order.Quantity]);
+                        } else if (type === "sell") {
+                            orderBooks[index].asks.push([order.Rate, order.Quantity]);
+                        }
+                    }
+
+                    for(let orderBook of orderBooks) {
```
Comment (kroitor) [inline js/ccex.js]: `for .. of ..`  constructs are not supported yet, we have to unfold them to basic counter-loops...

**Comment by pmizel** [comment]
ccex looks ok, i can't interprate this transpile error.

'1' of null ?

File: js/cryptopia.js [hunk 4]
```diff
@@ -169,6 +171,44 @@ module.exports = class cryptopia extends Exchange {
...
+    async fetchOrderBooks(symbols, params = {}){
+        await this.loadMarkets ();
+        let requestSymbols = [];
+        let orderBooks= [];
+        if(!symbols) {
+            symbols = this.symbols;
+        }
+        for (let i = 0; i < symbols.length; i++) {
+            requestSymbols.push(symbols[i]);
+            if ((i % 200 === 0 && i > 0) || (i === symbols.length - 1)) {
```
Comment (kroitor) [inline js/cryptopia.js]: This conditional is not portable (won't transpile to Python/PHP)

File: js/ccex.js (hunk 1 above)
Comment (pmizel) [reply js/ccex.js]: i try to find any empty lines and to have a single line between methods.

**Comment by kroitor** [comment]
@pmizel 

> ccex looks ok, i can't interprate this transpile error.
> '1' of null ?

Basically, we have to check each line of code against these rules: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes

ccex

**Comment by pmizel** [comment]
i try to fix it.

**Comment by kroitor** [comment]
@pmizel thank you so much! Really appreciate your involvement!

**Comment by pmizel** [comment]
i try to close it and start a new one.

**Comment by kroitor** [comment]
@pmizel don't worry too much if you can't defeat some transpilation errors – the transpiler is very sensitive, and it may complain at some non-intuitive things, but I will clean it up upon merging anyway. Thx again!

**Comment by pmizel** [comment]
ERROR: InvocationError: '/home/travis/build/ccxt/ccxt/python/.tox/qa/bin/flake8'
im not an python expert. sorry.

**Comment by kroitor** [comment]
Yeah, it's mostly because of minor syntax issues, have to clean it up further for transpiling:

```
./python/ccxt/liqui.py:1:1: E902 TokenError: EOF in multi-line statement
./python/ccxt/liqui.py:209:25: E225 missing whitespace around operator
./python/ccxt/liqui.py:210:26: E999 SyntaxError: invalid syntax
./python/ccxt/exmo.py:140:51: E225 missing whitespace around operator
./python/ccxt/exmo.py:140:51: E999 SyntaxError: invalid syntax
./python/ccxt/exmo.py:140:52: E251 unexpected spaces around keyword / parameter equals
./python/ccxt/exmo.py:141:31: E225 missing whitespace around operator
./python/ccxt/exmo.py:141:32: E251 unexpected spaces around keyword / parameter equals
./python/ccxt/exmo.py:155:21: E113 unexpected indentation
./python/ccxt/exmo.py:155:90: E225 missing whitespace around operator
./python/ccxt/cryptopia.py:1:1: E902 TokenError: EOF in multi-line statement
./python/ccxt/cryptopia.py:172:9: E265 block comment should start with '# '
./python/ccxt/cryptopia.py:175:26: E999 SyntaxError: invalid syntax
./python/ccxt/ccex.py:1:1: E902 TokenError: EOF in multi-line statement
./python/ccxt/ccex.py:133:68: E999 SyntaxError: invalid syntax
./python/ccxt/async/liqui.py:1:1: E902 TokenError: EOF in multi-line statement
./python/ccxt/async/liqui.py:209:25: E225 missing whitespace around operator
./python/ccxt/async/liqui.py:210:26: E999 SyntaxError: invalid syntax
./python/ccxt/async/exmo.py:140:51: E225 missing whitespace around operator
./python/ccxt/async/exmo.py:140:51: E999 SyntaxError: invalid syntax
./python/ccxt/async/exmo.py:140:52: E251 unexpected spaces around keyword / parameter equals
./python/ccxt/async/exmo.py:141:31: E225 missing whitespace around operator
./python/ccxt/async/exmo.py:141:32: E251 unexpected spaces around keyword / parameter equals
./python/ccxt/async/exmo.py:155:21: E113 unexpected indentation
./python/ccxt/async/exmo.py:155:90: E225 missing whitespace around operator
./python/ccxt/async/cryptopia.py:1:1: E902 TokenError: EOF in multi-line statement
./python/ccxt/async/cryptopia.py:172:9: E265 block comment should start with '# '
./python/ccxt/async/cryptopia.py:175:26: E999 SyntaxError: invalid syntax
./python/ccxt/async/ccex.py:1:1: E902 TokenError: EOF in multi-line statement
./python/ccxt/async/ccex.py:133:68: E999 SyntaxError: invalid syntax
```

But, if you're not comfortable with testing the build locally, don't worry, I'll edit it and will merge it anyway. Hope to do it asap. Thanks again for your awesome help!

**Comment by pmizel** [comment]
Thank you. Im out for today.

**Comment by pmizel** [comment]
Привет Игорь, ты знае когда у тебя будет время для интегразий #1033?

**Comment by kroitor** [comment]
@pmizel привет! Я постараюсь в ближайшие дни, но у меня очень много задач, поэтому я не могу обещать это сегодня или завтра, но я это сделаю в скором времени так или иначе. Спасибо!

**Comment by gschmuck** [comment]
Hey Igor, 

when will you merge this?

Greetings 
Gamal

**Comment by kroitor** [comment]
> when will you merge this?

It's on my list of closest tasks, however, it requires a significant rework to make it pass the tests, so I'll spend at least day on it, among other tasks.

**Comment by gschmuck** [comment]
Thanks for your effort! :)

**Comment by Codewyvern** [comment]
I just started using this library and I came across this issue, so let's say I am building a coin info app, I query the currencies first, get a list back, then query each pair, for each currency I check if it has a BTC pairing and then I call fetchOrderBook to get the current market price. Am I doing it correctly? doesn't this mean that I have to wait 1 second between each fetchOrderBook call? This is why I found this issue, wouldn't it be great to just provide a list of pairs and fetch the current market price for each one?

**Pull request merged**